File: app/src/main/java/com/inqint/yarnrequirements/ProjectFragment.java
package com.inqint.yarnrequirements;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.GridLayout;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.TextView;

import com.inqint.yarnrequirements.Projects.GaugeUnits;
import com.inqint.yarnrequirements.Projects.LongLengthUnits;
import com.inqint.yarnrequirements.Projects.Project;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;


/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link ProjectFragment.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link ProjectFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class ProjectFragment extends Fragment implements AdapterView.OnItemSelectedListener {
    private static final String ARG_PROJECT = "project";

    protected Project project;

    private OnFragmentInteractionListener mListener;

    public ProjectFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param project Parameter 1.
     * @return A new instance of fragment ProjectFragment.
     */
    public static ProjectFragment newInstance(Project project) {
        ProjectFragment fragment = new ProjectFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PROJECT, project.getName());
        fragment.setArguments(args);
        fragment.project = project;
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            project = ProjectContent.PROJECT_MAP.get(savedInstanceState.getString(ARG_PROJECT));
        }
    }

    protected View view;  // parent view
    protected GridLayout gridLayout;
    protected Context context;
    protected SharedPreferences preferences;
    private TextView name;
    private ImageView image;
    private EditText gaugeText;
    private Spinner gaugeUnits;
    protected TextView yarnNeeded;
    protected Spinner yarnUnits;
    protected EditText ballSize;
    protected Spinner ballUnits;
    protected TextView ballsNeeded;
    protected Spinner partialBalls;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        view = inflater.inflate(R.layout.fragment_project, container, false);
        context = view.getContext();
        preferences = context.getSharedPreferences(project.getName(), Context.MODE_PRIVATE);

        String filename = String.format(view.getResources().getString(R.string.project_settings),
                project.getName().toLowerCase());
        project.getSettings(preferences, readDefaults(filename));

        initValues();
        initGaugeUnitSpinner();
        initYarnUnitSpinner();
        initBallUnitSpinner();
        initPartialBallSpinner();

        initTextChangedEvents();

        name.setText(project.getName());
        image.setImageResource(project.getImageID());
        setInitialValues();

        return view;
    }
    // Read the default settings for the project
    protected JSONObject readDefaults(String file) {
        JSONObject object = null;
        try {
            InputStream is  = getActivity().getAssets().open(file);
            int size = is.available();
            byte[] buffer = new byte[size];
            is.read(buffer);
            is.close();
            String json = new String(buffer, "UTF-8");
            object = new JSONObject(json);
        } catch (IOException e) {
            e.printStackTrace();
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return object;
    }
    // Identify the various views
    protected void initValues() {
        name = (TextView) view.findViewById(R.id.name);
        image = (ImageView) view.findViewById(R.id.image);
        gridLayout = (GridLayout) view.findViewById(R.id.gridLayout);
        gaugeText = (EditText) view.findViewById(R.id.editGauge);
        gaugeUnits = (Spinner) view.findViewById(R.id.gaugeUnitsSpinner);
        yarnNeeded = (TextView) view.findViewById(R.id.yarnNeededText);
        yarnUnits = (Spinner) view.findViewById(R.id.yarnUnitsSpinner);
        ballSize = (EditText) view.findViewById(R.id.editBallSize);
        ballUnits = (Spinner) view.findViewById(R.id.ballSizeSpinner);
        ballsNeeded = (TextView) view.findViewById(R.id.ballsNeededText);
        partialBalls = (Spinner) view.findViewById(R.id.ballFractSpinner);
    }
    // Set the values for all the views
    protected void setInitialValues() {
        gaugeText.setText(Double.toString(project.getGauge()));
        gaugeUnits.setSelection(project.getGaugeUnits().ordinal());
        yarnNeeded.setText(Integer.toString(project.getYarnNeeded()));
        yarnUnits.setSelection(project.getYarnNeededUnits().ordinal());
        ballSize.setText(Integer.toString(project.getBallSize()));
        ballUnits.setSelection(project.getBallSizeUnits().ordinal());
        ballsNeeded.setText(Double.toString(project.getBallsNeeded()));
        partialBalls.setSelection(project.isPartialBalls() ? 1 : 0);
    }

    // Set up the Gauge Unit spinner
    private void initGaugeUnitSpinner() {
        // Create an ArrayAdapter using the string array and a default spinner layout
        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(view.getContext(),
                R.array.gauge_units_array, android.R.layout.simple_spinner_item);
        // Specify the layout to use when the list of choices appears
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        // Apply the adapter to the spinner
        gaugeUnits.setAdapter(adapter);

        gaugeUnits.setOnItemSelectedListener(this);
    }

    private void initYarnUnitSpinner() {
        // Create an ArrayAdapter using the string array and a default spinner layout
        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(context,
                R.array.long_length_units_array, android.R.layout.simple_spinner_item);
        // Specify the layout to use when the list of choices appears
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        // Apply the adapter to the spinner
        yarnUnits.setAdapter(adapter);

        yarnUnits.setOnItemSelectedListener(this);
    }

    private void initBallUnitSpinner() {
        // Create an ArrayAdapter using the string array and a default spinner layout
        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(context,
                R.array.long_length_units_array, android.R.layout.simple_spinner_item);
        // Specify the layout to use when the list of choices appears
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        // Apply the adapter to the spinner
        ballUnits.setAdapter(adapter);

        ballUnits.setOnItemSelectedListener(this);
    }

    private void initPartialBallSpinner() {
        // Create an ArrayAdapter using the string array and a default spinner layout
        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(context,
                R.array.partial_ball_array, android.R.layout.simple_spinner_item);
        // Specify the layout to use when the list of choices appears
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        // Apply the adapter to the spinner
        partialBalls.setAdapter(adapter);

        partialBalls.setOnItemSelectedListener(this);
    }

    protected void initTextChangedEvents() {
        gaugeText.addTextChangedListener(new TextWatcher() {
            public void afterTextChanged(Editable s) {
                String str = gaugeText.getText().toString();
                if (str.length() > 0) {
                    project.setGauge(Double.parseDouble(str));
                    project.calcYarnRequired();
                    updateResults();
                }
            }

            public void beforeTextChanged(CharSequence arg0, int arg1,
                                          int arg2, int arg3) {
                //  Auto-generated method stub

            }

            public void onTextChanged(CharSequence s, int start, int before,
                                      int count) {
                //  Auto-generated method stub
            }
        });

        final EditText ballSizeText = (EditText) view.findViewById(R.id.editBallSize);
        ballSizeText.addTextChangedListener(new TextWatcher() {
            public void afterTextChanged(Editable s) {
                String str = ballSizeText.getText().toString();
                if (str.length() > 0) {
                    project.setBallSize(Integer.parseInt(str));
                    project.calcBallsNeeded();
                    updateResults();
                }
            }

            public void beforeTextChanged(CharSequence arg0, int arg1,
                                          int arg2, int arg3) {
                //  Auto-generated method stub

            }

            public void onTextChanged(CharSequence s, int start, int before,
                                      int count) {
                //  Auto-generated method stub
            }
        });
    }

    public void updateResults() {
        TextView yarnNeeded = (TextView) view.findViewById(R.id.yarnNeededText);
        TextView ballsNeeded = (TextView) view.findViewById(R.id.ballsNeededText);

        yarnNeeded.setText(String.format("%d", project.getYarnNeeded()));
        ballsNeeded.setText(String.format("%.1f", project.getBallsNeeded()));
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int pos, long id) {
        switch (parent.getId()) {
            case R.id.gaugeUnitsSpinner:
                project.setGaugeUnits(GaugeUnits.fromInt(pos));
                project.calcYarnRequired();
                break;
            case R.id.yarnUnitsSpinner:
                project.setYarnNeededUnits(LongLengthUnits.fromInt(pos));
                break;
            case R.id.ballSizeSpinner:
                project.setBallSizeUnits(LongLengthUnits.fromInt(pos));
                project.calcBallsNeeded();
                break;
            case R.id.ballFractSpinner:
                project.setPartialBalls(pos != 0);
                break;
        }
        updateResults();
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {
    }

    public void setProject(Project project) {
        this.project = project;
    }


    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        // TODO: Update argument type and name
        void onFragmentInteraction(Uri uri);
    }
}

File: app/src/main/java/com/inqint/yarnrequirements/ProjectFragment.kt
package com.inqint.yarnrequirements

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.os.Parcelable
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import com.inqint.yarnrequirements.Projects.GaugeUnits
import com.inqint.yarnrequirements.Projects.LongLengthUnits
import com.inqint.yarnrequirements.Projects.Project
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.io.Serializable
import java.nio.charset.Charset


/**
 * A simple [Fragment] subclass.
 * Activities that contain this fragment must implement the
 * [ProjectFragment.OnFragmentInteractionListener] interface
 * to handle interaction events.
 * Use the [ProjectFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
open class ProjectFragment : Fragment(), AdapterView.OnItemSelectedListener {

    //private var listener: OnFragmentInteractionListener? = null
    lateinit var project: Project
    protected lateinit var gridLayout: GridLayout
    protected lateinit var preferences: SharedPreferences
    private lateinit var name: TextView
    private lateinit var image: ImageView
    private lateinit var gaugeText: EditText
    private lateinit var gaugeUnits: Spinner
    protected lateinit var yarnNeeded: TextView
    protected lateinit var yarnUnits: Spinner
    protected lateinit var ballSize: EditText
    protected lateinit var ballUnits: Spinner
    protected lateinit var ballsNeeded: TextView
    protected lateinit var partialBalls: Spinner
    protected lateinit var mview: View
    protected var userUpdate: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            val bundle = arguments!!
            if (bundle.containsKey("project")) {
                if (Parcelable::class.java.isAssignableFrom(Project::class.java) ||
                    Serializable::class.java.isAssignableFrom(Project::class.java)) {
                    project = bundle.get("project") as Project
                } else {
                    throw UnsupportedOperationException(Project::class.java.name +
                            " must implement Parcelable or Serializable or must be an Enum.")
                }
                if (project == null) {
                    throw IllegalArgumentException("Argument \"project\" is marked as non-null but was passed a null value.")
                }
            } else {
                throw IllegalArgumentException("Required argument \"project\" is missing and does not have an android:defaultValue")
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        mview = inflater.inflate(R.layout.fragment_project, container, false)
        val context = mview.context
        preferences = context.getSharedPreferences(project.name, Context.MODE_PRIVATE)

        val filename = String.format(
            mview.resources.getString(R.string.project_settings),
            project.name.toLowerCase()
        )
        project.getSettings(preferences, readDefaults(filename)!!)

        initValues()
        initGaugeUnitSpinner()
        initYarnUnitSpinner()
        initBallUnitSpinner()
        initPartialBallSpinner()

        initTextChangedEvents()

        name.text = project.name
        image.setImageResource(project.imageID)
        setInitialValues()

        return mview
    }

    // Read the default settings for the project
    protected fun readDefaults(file: String): JSONObject? {
        var `object`: JSONObject? = null
        try {
            val `is` = activity!!.assets.open(file)
            val size = `is`.available()
            val buffer = ByteArray(size)
            `is`.read(buffer)
            `is`.close()
            val json = String(buffer, Charset.forName("UTF-8"))
            `object` = JSONObject(json)
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        return `object`
    }

    // Identify the various views
    protected open fun initValues() {
        name = mview.findViewById<View>(R.id.name) as TextView
        image = mview.findViewById<View>(R.id.image) as ImageView
        gridLayout = mview.findViewById(R.id.gridLayout) as GridLayout
        gaugeText = mview.findViewById(R.id.editGauge) as EditText
        gaugeUnits = mview.findViewById(R.id.gaugeUnitsSpinner) as Spinner
        yarnNeeded = mview.findViewById(R.id.yarnNeededText) as TextView
        yarnUnits = mview.findViewById(R.id.yarnUnitsSpinner) as Spinner
        ballSize = mview.findViewById(R.id.editBallSize) as EditText
        ballUnits = mview.findViewById(R.id.ballSizeSpinner) as Spinner
        ballsNeeded = mview.findViewById(R.id.ballsNeededText) as TextView
        partialBalls = mview.findViewById(R.id.ballFractSpinner) as Spinner
    }

    // Set the values for all the views
    protected fun setInitialValues() {
        gaugeText.setText(String.format("%.1f", project.gauge))
        gaugeUnits.setSelection(project.gaugeUnits.ordinal)
        yarnNeeded.text = String.format("%d", project.yarnNeeded)
        yarnUnits.setSelection(project.yarnNeededUnits.ordinal)
        ballSize.setText(String.format("%d", project.ballSize))
        ballUnits.setSelection(project.ballSizeUnits.ordinal)
        ballsNeeded.text = String.format("%.1f", project.ballsNeeded)
        partialBalls.setSelection(if (project.isPartialBalls) 1 else 0)
    }

    // Set up the Gauge Unit spinner
    private fun initGaugeUnitSpinner() {
        // Create an ArrayAdapter using the string array and a default spinner layout
        val adapter = ArrayAdapter.createFromResource(
            mview.context,
            R.array.gauge_units_array, R.layout.spinner
        )
        // Specify the layout to use when the list of choices appears
        adapter.setDropDownViewResource(R.layout.spinner)
        // Apply the adapter to the spinner
        gaugeUnits.adapter = adapter

        gaugeUnits.onItemSelectedListener = this
    }

    private fun initYarnUnitSpinner() {
        // Create an ArrayAdapter using the string array and a default spinner layout
        val adapter = ArrayAdapter.createFromResource(
            context!!,
            R.array.long_length_units_array, R.layout.spinner
        )
        // Specify the layout to use when the list of choices appears
        adapter.setDropDownViewResource(R.layout.spinner)
        // Apply the adapter to the spinner
        yarnUnits.adapter = adapter

        yarnUnits.onItemSelectedListener = this
    }

    private fun initBallUnitSpinner() {
        // Create an ArrayAdapter using the string array and a default spinner layout
        val adapter = ArrayAdapter.createFromResource(
            context,
            R.array.long_length_units_array, R.layout.spinner
        )
        // Specify the layout to use when the list of choices appears
        adapter.setDropDownViewResource(R.layout.spinner)
        // Apply the adapter to the spinner
        ballUnits.adapter = adapter

        ballUnits.onItemSelectedListener = this
    }

    private fun initPartialBallSpinner() {
        // Create an ArrayAdapter using the string array and a default spinner layout
        val adapter = ArrayAdapter.createFromResource(
            context,
            R.array.partial_ball_array, R.layout.spinner
        )
        // Specify the layout to use when the list of choices appears
        adapter.setDropDownViewResource(R.layout.spinner)
        // Apply the adapter to the spinner
        partialBalls.adapter = adapter

        partialBalls.onItemSelectedListener = this
    }

    protected open fun initTextChangedEvents() {
        gaugeText.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable) {
                val str = gaugeText.text.toString()
                if (userUpdate && str.isNotEmpty()) {
                    project.gauge = java.lang.Double.parseDouble(str)
                    with (preferences.edit()) {
                        putFloat("gauge", project.gauge.toFloat())
                        commit()
                    }
                    project.calcYarnRequired()
                    updateResults()
                }
            }

            override fun beforeTextChanged(
                arg0: CharSequence, arg1: Int,
                arg2: Int, arg3: Int
            ) {
                //  Auto-generated method stub

            }

            override fun onTextChanged(
                s: CharSequence, start: Int, before: Int,
                count: Int
            ) {
                //  Auto-generated method stub
            }
        })

        val ballSizeText = mview.findViewById(R.id.editBallSize) as EditText
        ballSizeText.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable) {
                val str = ballSizeText.text.toString()
                if (str.isNotEmpty()) {
                    project.ballSize = Integer.parseInt(str)
                    with (preferences.edit()) {
                        putInt("ballSize", project.ballSize)
                        commit()
                    }
                    project.calcBallsNeeded()
                    updateResults()
                }
            }

            override fun beforeTextChanged(
                arg0: CharSequence, arg1: Int,
                arg2: Int, arg3: Int
            ) {
                //  Auto-generated method stub

            }

            override fun onTextChanged(
                s: CharSequence, start: Int, before: Int,
                count: Int
            ) {
                //  Auto-generated method stub
            }
        })
    }

    open fun updateResults() {
        userUpdate = false
        val yarnNeeded = mview.findViewById(R.id.yarnNeededText) as TextView
        val ballsNeeded = mview.findViewById(R.id.ballsNeededText) as TextView

        yarnNeeded.text = String.format("%d", project.yarnNeeded)
        ballsNeeded.text = String.format("%.1f", project.ballsNeeded)
        userUpdate = true
    }

    override fun onItemSelected(parent: AdapterView<*>, view: View, pos: Int, id: Long) {
        when (parent.id) {
            R.id.gaugeUnitsSpinner -> {
                project.gaugeUnits = GaugeUnits.fromInt(pos)
                with (preferences.edit()) {
                    putInt("gaugeUnits", pos)
                    commit()
                }
                project.calcYarnRequired()
            }
            R.id.yarnUnitsSpinner -> {
                project.yarnNeededUnits = LongLengthUnits.fromInt(pos)
                with (preferences.edit()) {
                    putInt("yarnNeededUnits", pos)
                    commit()
                }
                project.calcYarnRequired()
            }
            R.id.ballSizeSpinner -> {
                project.ballSizeUnits = LongLengthUnits.fromInt(pos)
                with (preferences.edit()) {
                    putInt("ballSizeUnits", pos)
                    commit()
                }
                project.calcBallsNeeded()
            }
            R.id.ballFractSpinner -> {
                project.isPartialBalls = pos != 0
                project.calcBallsNeeded()
                with (preferences.edit()) {
                    putBoolean("isPartialBalls", project.isPartialBalls)
                    commit()
                }
            }
        }
        updateResults()
    }

    override fun onNothingSelected(parent: AdapterView<*>) {}

    companion object {
        private const val ARG_PROJECT = "project"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param project Parameter 1.
         * @return A new instance of fragment ProjectFragment.
         */
        fun newInstance(pair: ProjectContent.FragmentProject): ProjectFragment {
            val args = Bundle()
            args.putSerializable(ARG_PROJECT, pair.project)
            val ctor = pair.fragment.constructors.singleOrNull()!! as () -> ProjectFragment
            val fragment = ctor()
            fragment.arguments = args
            return fragment
        }
    }
}
